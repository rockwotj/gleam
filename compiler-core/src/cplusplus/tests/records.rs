use crate::assert_cpp;

#[test]
fn single_variant() {
    assert_cpp!(
        r#"
pub type Person { Person(name: String, age: Int) }
"#,
    );
}

#[test]
fn unnamed_fields() {
    assert_cpp!(
        r#"
pub type Ip { Ip(String) }
"#,
    );
}

#[test]
fn multiple_variants() {
    assert_cpp!(
        r#"
pub type User {
  LoggedIn(name: String)
  Guest
}
"#,
    );
}

#[test]
fn shared_fields() {
    assert_cpp!(
        r#"
pub type Animal {
  Cat(name: String, likes_milk: Bool)
  Dog(name: String, barks: Bool)
}
"#,
    );
}

#[test]
fn generic_single_variant() {
    assert_cpp!(
        r#"
pub type Box(inner_type) {
  Box(inner: inner_type)
}
"#,
    );
}

#[test]
fn generic_multiple_variant() {
    assert_cpp!(
        r#"
pub type Either(left_type, right_type) {
  Left(v: left_type)
  Right(v: right_type)
}
"#,
    );
}

#[test]
fn singleton() {
    assert_cpp!(
        r#"
pub type Void { Void }

fn noop() -> Void {
  let v = Void;
  v
}
"#,
    );
}

#[test]
fn constructor_fn() {
    assert_cpp!(
        r#"
pub type Person { Person(name: String) }

fn new(name: String) -> Person {
  let constructor_fn = Person;
  constructor_fn(name)
}
"#,
    );
}
