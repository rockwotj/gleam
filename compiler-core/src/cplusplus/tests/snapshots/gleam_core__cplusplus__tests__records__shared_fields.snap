---
source: compiler-core/src/cplusplus/tests/records.rs
expression: "\npub type Animal {\n  Cat(name: String, likes_milk: Bool)\n  Dog(name: String, barks: Bool)\n}\n\npub fn make_german_shepard(name) {\n  let barks = True;\n  Dog(name, barks)\n}\n"
---
#ifndef MY_PACKAGE_MY_MODULE_H_
#define MY_PACKAGE_MY_MODULE_H_

#include <gleam.h>

namespace my {
namespace module {

struct Animal;
struct Animal$Cat;
struct Animal$Dog;

gleam::Ref<::my::module::Animal> make_german_shepard(gleam::String name);

struct Animal {
  explicit Animal(gleam::String name) : name(name) {}
  virtual ~Animal() = default;
  gleam::String name;
};

struct Cat$Animal : public Animal {
  explicit Cat$Animal(gleam::String name, bool likes_milk) : Animal(name), likes_milk(likes_milk) {}
  
  bool likes_milk;
};

struct Dog$Animal : public Animal {
  explicit Dog$Animal(gleam::String name, bool barks) : Animal(name), barks(barks) {}
  
  bool barks;
};

} // namespace my
} // namespace module

#endif // MY_PACKAGE_MY_MODULE_H_

---
#include "module.h"

namespace my {
namespace module {

gleam::Ref<::my::module::Animal> make_german_shepard(gleam::String name) {
  bool barks = true;
  return gleam::MakeRef<::my::module::Animal$Dog>(name, barks);
};

} // namespace my
} // namespace module

